security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -374,21 +374,29 @@ func NotifyProxy(conID, resourceName string, secret *model.SecretItem) error {
 }
 
 func recycleConnection(conID, resourceName string) {
-	metricLabelName := generateResourcePerConnLabel(resourceName, conID)
-	sdsMetrics.pushPerConn.DeleteLabelValues(metricLabelName)
-	sdsMetrics.pendingPushPerConn.DeleteLabelValues(metricLabelName)
-	sdsMetrics.pushErrorPerConn.DeleteLabelValues(metricLabelName)
-	sdsMetrics.staleConn.WithLabelValues(metricLabelName).Inc()
-	sdsMetrics.totalStaleConn.Inc()
-	sdsMetrics.totalActiveConn.Dec()
 	key := cache.ConnKey{
 		ConnectionID: conID,
 		ResourceName: resourceName,
 	}
 
+	// Only add connection key to staledClientKeys if it's not there already.
+	// The recycleConnection function may be triggered more than once for each connection key.
+	// https://github.com/istio/istio/issues/15306#issuecomment-509783105
+	if _, found := staledClientKeys[key]; found {
```

Comment:
move this after
```
defer sdsClientsMutex.Unlock()
```

Reply:
hmm should there be a race condition if we keep it this way? We simply read `staledClientKeys` map here.

Reply:
yes there is race condition there. ```recycleConnection``` could be called from different go-routine.

Reply:
Right :D